awswrangler/mysql.py

```diff
@@ -13,7 +13,13 @@
 from awswrangler import _databases as _db_utils
 from awswrangler._config import apply_configs
 
-pymysql = _utils.import_optional_dependency("pymysql")
+if TYPE_CHECKING:
+    try:
+        import pymysql
+    except ImportError:
+        pass
+else:
+    pymysql = _utils.import_optional_dependency("pymysql")
```

Comment:
Is the issue here that if the module is not in the path, mypy fails type checks? Would it help to add types to `import_optional_dependency` based on the literal input (similar to what was done with clients)? & perhaps move `if TYPE_CHECKING` inside the function or add it as input argument.

Reply:
Second this, can you clarify why this is required? The fix looks verbose so would be ideal if we can abstract it

Reply:
I've tried giving a bunch of overloads to `import_optional_dependency` so that it returns the proper module type. Unfortunately, this doesn't seem to be something `mypy` knows how to handle yet.

I had the following code in `_utils.py`::
```python
if TYPE_CHECKING:
    try:
        import pymysql
    except ImportError:
        pass

@overload
def import_optional_dependency(name: Literal["pymysql"]) -> "pymysql":
    ...

@overload
def import_optional_dependency(name: str) -> ModuleType:
    ...
```

which outputted the following errors:
```
awswrangler/_utils.py:112: error: Module "pymysql" is not valid as a type  [valid-type]
awswrangler/_utils.py:112: note: Perhaps you meant to use a protocol matching the module structure?
awswrangler/mysql.py:22: error: pymysql? has no attribute "connections"  [attr-defined]
awswrangler/mysql.py:161: error: pymysql? has no attribute "connect"  [attr-defined]
awswrangler/mysql.py:171: error: pymysql? has no attribute "cursors"  [attr-defined]
awswrangler/mysql.py:499: error: pymysql? has no attribute "cursors"  [attr-defined]
```